All things log-in!

ryan-garay89 opened the pull request:
This PR adds full functionality for logging in and out. I wrote this code fairly quickly so please make sure it looks okay. The login flow implemented here **does not** verify that a user is an admin or counselor, but rather assumes that the user is a student. We will have to implement functionality for verifying user types at a later time.

I ask you to please check out this branch locally and do some testing / play around with it and see if you find any bugs. Once it's ready to be merged we will update the redirect_uri on Google.

KLongLo committed:
need debug

ryan-garay89 committed:
update button:

ryan-garay89 committed:
switched to post

ryan-garay89 committed:
remove signup

ryan-garay89 committed:
button styling

ryan-garay89 committed:
auth workinggit status

ryan-garay89 committed:
add migration

ryan-garay89 committed:
final push

ryan-garay89 committed:
clean

ryan-garay89 committed:
assume user is student

ryan-garay89 commented on app/views/pages/index.html.erb:
```diff
@@ -9,24 +9,22 @@
         <%# pending %>
-      <% end %>
-      <% if @user_type.include? 'Student' %>
+      <% else %>
```
for now, at least before admins & counselors are fully functional, let's assume a user is a student if they have no other type

KLongLo commented on app/controllers/user_controller.rb:
```diff
@@ -0,0 +1,35 @@
+      redirect_to login_confirm_path, flash: { :error => "Something went wrong, please try again" }
+      return
+    end
+    if sid.length != 10
```
Here should check if sid is not nil, otherwise sid.length will raise an error

KLongLo replied:
Other things look good to me :)

ryan-garay89 replied:
will do! thank you

KLongLo commented on app/controllers/sessions_controller.rb:
```diff
@@ -4,14 +4,22 @@ def googleAuth
         user = User.from_omniauth(access_token)
-        log_in(user)
         # Access_token is used to authenticate request made from the rails application to the google server
         user.google_token = access_token.credentials.token
         # Refresh_token to request new access_token
         # Note: Refresh_token is only sent once during the first request
         refresh_token = access_token.credentials.refresh_token
         user.google_refresh_token = refresh_token if refresh_token.present?
```
Do we really need these lines?

ryan-garay89 replied:
I think they allow google to request new logins for security purposes (the google auth expires) -- maybe @michellekroll knows more?

michellekroll replied:
I am not sure but I can look into it. I believe the functionality we have right now will not do that though.

HaolanMo committed:
try login step def

ryan-garay89 committed:
update login, testing

ryan-garay89 commented on app/controllers/sessions_controller.rb:
```diff
@@ -3,15 +3,28 @@ class SessionsController < ApplicationController
+                redirect_to root_path, flash: { :success => "Success! You've been logged-in!" }
+            else
+                session[:current_user_id] = user.id
+                redirect_to login_confirm_path
```
only new users should be redirected to verify their information

KLongLo replied:
Should we do user.save only if existing_user does not exist?

ryan-garay89 commented on app/controllers/user_controller.rb:
```diff
@@ -0,0 +1,23 @@
+    last_name = params[:user][:last_name]
+    email = params[:user][:email]
+    sid = params[:user][:sid]
+    if sid.blank? or sid.length != 10
```
I updated the checks for first_name, last_name, etc. to be validated in the User model. Now I only need to check that the sid entered isn't blank (nil or empty) and that the length is valid.

ryan-garay89 commented on app/views/login/confirm.html.erb:
```diff
@@ -0,0 +1,41 @@
+          <div class="col-md-6 offset-md-3">
+            <div class="col-md-6 offset-md-3">
+              <%= f.label :first_name, class: "form-label" %>
+              <%= f.text_field :first_name, value: @user.first_name, :required => true, class: "form-control" %>
```
added :required fields to text inputs on the form to validate on the frontend that fields aren't empty

ryan-garay89 committed:
rename, test

KLongLo commented:
When the same user logs in the second time, the landing page will look like this, which doesn't look right 🤔:
![image](https://user-images.githubusercontent.com/56783069/162342931-d114a2b4-7e9c-4b0b-ac4f-69b9e7c449cf.png)


edit: I think the problem occurs because the program doesn't set `session[:current_session_id]` if the logging in user is an existing user?

KLongLo commented:
Also, I think it is better to write some integration tests for the login features before merging to the main branch?

ryan-garay89 commented:
> When the same user logs in the second time, the landing page will look like this, which doesn't look right 🤔:
> 
> ![image](https://user-images.githubusercontent.com/56783069/162342931-d114a2b4-7e9c-4b0b-ac4f-69b9e7c449cf.png)
> 
> 
> 
> 
> 
> edit: I think the problem occurs because the program doesn't set `session[:current_session_id]` if the logging in user is an existing user?

yes, my bad! I will fix

KLongLo commented:
The confirm.html.erb is actually updating the attributes of an user. Should make a new method in user controller (and its corresponding view) for handling the attribute updates?

ryan-garay89 commented:
> The confirm.html.erb is actually updating the attributes of an user. Should make a new method in user controller (and its corresponding view) for handling the attribute updates?

the confirm.html.erb doesn't actually change or interact with the record as far as I know. And the attribute updates are handled in the user controller (in the update method)

HaolanMo committed:
finish some login testing and step definitions

HaolanMo committed:
fix some syntax error

HaolanMo committed:
fix some syntax error v2

HaolanMo committed:
fix some syntax error v3

HaolanMo committed:
fix some syntax error v4

HaolanMo committed:
fix some syntax error v5

ryan-garay89 committed:
fix known login:

ryan-garay89 committed:
Merge branch 'main' into 181301090-login

ryan-garay89 committed:
styling

ryan-garay89 committed:
merged

ryan-garay89 committed:
plural users

ryan-garay89 committed:
html_safe & styling flash

KLongLo committed:
test: finish tests

Scenario: student should not be able to check-in without filling in a reason (sad path) failed because capybara doesn't block user from submitting checkin without reason

KLongLo committed:
fix: remove byebug

KLongLo committed:
fix: change steps for a scenario

Scenario: student should not be able to check-in without filling in a reason (sad path)

KLongLo committed:
Update main.yml

change command for running unit tests and cucumber tests

KLongLo committed:
refactor: separate sad_path.feature to multiple features

KLongLo committed:
Merge branch 'cucumber-testing-based-on-login' of https://github.com/ryan-garay89/berkeley-reentry-student-program into cucumber-testing-based-on-login

KLongLo committed:
Merge branch '181301090-login' into cucumber-testing-based-on-login

ryan-garay89 committed:
existing_user refactor

ryan-garay89 committed:
fix

KLongLo committed:
Merge pull request #34 from ryan-garay89/cucumber-testing-based-on-login

Cucumber testing based on login

ryan-garay89 committed:
edit workflow

KLongLo reviewed (approved):
LGTM!!

ryan-garay89 committed:
conventional update
